src/RequestFactoryInterface.php

```diff
@@ -5,15 +5,19 @@
 use Psr\Http\Message\RequestInterface;
 use Psr\Http\Message\UriInterface;
 
-interface RequestFactoryInterface
+interface RequestFactoryInterface extends
+    StreamFactoryInterface,
+    UriFactoryInterface
 {
     /**
      * Create a new request.
      *
-     * @param string $method
-     * @param UriInterface|string $uri
+     * @param string $method The HTTP method associated with the request.
+     * @param UriInterface|string $uri The URI associated with the request. If
+     *     the value is a string, the factory MUST create a UriInterface
+     *     instance based on it.
      *
      * @return RequestInterface
      */
-    public function createRequest($method, $uri);
+    public function createRequest(string $method, $uri): RequestInterface;
```

Comment:
I think we should make `$uri` strictly typed to `UriInterface`, because this is easy:

```php
$requestFactory->createRequest('GET', $requestFactory->createUri('http://example.com'));
```

now that `RequestFactoryInterface` extends `UriFactoryInterface`

Reply:
I'm actually a bit less sure about this one - compared with `UploadedFileFactoryInterface`, which I commented on regarding the same issue below, I think this one is different, because the URI factory has only one method. There's only way to create a URI, and that's from a string. We lose nothing by allowing `string` here and it's not competing with `UriInterface` for any functionality - the URI factory provides no additional options.

I think we can safely favor brevity in this case, and I think it makes sense to do so, since the request-factory is otherwise literally useless on it's own; using the request-factory would come with a hard dependency on the URI factory, which doesn't really make sense, when there's no advantage for the majority use-case. (marginal use-cases can of course still use the URI factory, if needed, but it's hard to think of a use-case.)

Reply:
I totally agree with @mindplay-dk here. Forcing the user to use the factory to also create the URI makes the request factory far less usable, for no good reason. Since the RequestFactory extends the UriFactory in this case, the `$uri` value can be passed on directly to the `createUri()` method in the case of a string.